### DISCUSSION TITLE
Fileupload in a repeater

### DISCUSSION THREAD

**Discussion by horneberh** [filamentphp/filament#1931] [Help] [answered]
Hi there,
I want to use a fileuploader field with multiple uploads enabled in a repeater field ( we need a gallery with title and n images). Furthermore I am using the form builder as a standalone module in my app. First question: is this possible at all? Because I have two problems now:
- The repeater field gets saved in a json field. That's great. But I can't cast this field back to read it. When I do I get the error, that the BasicFileUpload expects a string and not an array.
- Although I set the disk to save on my public disk, the json data states, that the images got saved on my "local" disk (which they didn't).

Perhaps you guys have an idea, how I could solve this. Thx!

**Comment by danharrin** [comment, marked as answer]
It's definitely possible, please send the code you tried. Not sure why they are saved to your local disk, Filament deliberately sets your public disk as default. Make sure your public storage is linked,

**Reply by horneberh** [reply]
That sounds good. So, the database field contains this:
`{"4ebb1aa6-d105-4bec-9029-c84a120d9f86": {"title": "Test", "images": {"427a7977-8072-4017-a141-4fe97203a8ff": {"disk": "local"}, "9e4a784c-f4d4-41eb-9b31-44657f61355f": {"disk": "local"}, "f4f0276a-2492-49cc-a655-2e88fc7f0de3": {"disk": "local"}}}}`
And I did check if the storage is linked and the directory to save the images is exists too.

This is the field:
`Forms\Components\Repeater::make('galleries')
->schema([
    Forms\Components\TextInput::make('title')->required(),
    Forms\Components\FileUpload::make('images')
        ->image()
        ->disk('public')
        ->directory('bilder')
        ->visibility('public')
        ->multiple()
        ->enableReordering()
        ->preserveFilenames()
])
->columnSpan(6)`

This is the cast:
`    protected $casts = [
        'galleries'              => 'array',
    ];
`

The error I receive after casting is:
`Filament\Forms\Components\BaseFileUpload::Filament\Forms\Components\{closure}(): Argument #1 ($file) must be of type string, array given, called in .../vendor/laravel/framework/src/Illuminate/Collections/Collection.php on line 774`

Please tell me if you need other infos. Btw. filament is such a joy to work with. Thank you guys for this!

**Reply by danharrin** [reply]
Make sure you get the data from the repeater (for saving) using `$this->form->getState()`, dont save directly from `$this->galleries`.

I can see that you have objects like `{"disk": "local"}` in the saved data which is incorrect, if the form was saved correctly then this would be an array of file URLs instead.

**Reply by horneberh** [reply]
That did it! I am sorry, this is quite clear in the docs and I overlooked it. Thanks for your help, Dan.

**Reply by danharrin** [reply]
No worries! Glad you got it working mate.
